### PR TITLE
Set default branch to master if not set from git

### PR DESCRIPTION
**Pull request by sima-zhu**
Two changes in PR:
1. Set default branch to master if not set from git command. Add `continueOnError: true` to unblock any failed places.
E.g. https://dev.azure.com/azure-sdk/internal/_build/results?buildId=838864&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=bdeefc16-b669-5ebd-ad94-a2c19ade53b0

Testing pipeline right here: https://dev.azure.com/azure-sdk/playground/_build/results?buildId=845015&view=logs&j=38d918fc-df5a-5870-2b80-6c17fdb13baf&t=489baed9-6f06-5266-9d0d-079a130738b4
2. `git-push-changes.yml` checked whether local has changes. Make the changes so if we need to skip checking changes somehow, we are able to execute the scripts. 
E.g: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=842830&view=logs&j=321fe29c-d6e3-5c98-996f-e7c628f12971&t=71fd3457-22c6-5475-4ae6-23dcf07ca155

### CODE REVIEW

File: eng/common/pipelines/templates/steps/git-push-changes.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ steps:
 
 - task: PowerShell@2
   displayName: Push changes
-  condition: and(succeeded(), eq(variables['HasChanges'], 'true'))
+  condition: and(succeeded(), ne(variables['HasChanges'], 'false'))
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/git-push-changes.yml:44]: Why this change? We want to make sure we don't accidently start pushing these changes unless someone has explicitly set it to true.

File: eng/common/pipelines/templates/steps/set-default-branch.yml [hunk 2]
```diff
@@ -4,7 +4,12 @@ parameters:
...
     $setDefaultBranch = (git remote show ${{ parameters.RemoteRepo }} | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
+    if ($LASTEXITCODE -ne 0) {
+      Write-Host "Not able to fetch the default branch from git command. Set to master."
+      $setDefaultBranch = 'master'
+    }
     Write-Host "Setting DefaultBranch=$setDefaultBranch"
     echo "##vso[task.setvariable variable=DefaultBranch]$setDefaultBranch"
   displayName: "Setup Default Branch"
   workingDirectory: ${{ parameters.workingDirectory }}
+  continueOnError: true
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/set-default-branch.yml:15]: This will end up turning the pipeline yellow which we should try to avoid. Please do a test build on a private pipeline with this change and lets see how we can make this silent (i.e. no warnings/errors) only the write-host message.

File: eng/common/pipelines/templates/steps/set-default-branch.yml [hunk 3]
```diff
@@ -4,7 +4,12 @@ parameters:
 steps:
 - pwsh: |
     $setDefaultBranch = (git remote show ${{ parameters.RemoteRepo }} | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
+    if ($LASTEXITCODE -ne 0) {
+      Write-Host "Not able to fetch the default branch from git command. Set to master."
+      $setDefaultBranch = 'master'
+    }
     Write-Host "Setting DefaultBranch=$setDefaultBranch"
     echo "##vso[task.setvariable variable=DefaultBranch]$setDefaultBranch"
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/set-default-branch.yml:12]: nit: echo is just an alias for Write-Host and given you are using Write-Host in other places it would be nice to use it here as well.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 1 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:44]: Several cases here:
1. Step `Check for changes` executes:
* `git diff` success: Then HasChange=true, we run `git push` on purpose. 
* `git diff` fail: Then HasChange=false, we skip `git push` on purpose. 
2. Step `Check for changes` skips, then HasChanges not set, we will run `git push` step anyway:
* If people intended to check in some commits to branch, then they have to literally put PushArgs: -f. (_This is the purpose of the changes. In my case, I fetch some commits in master, and then merge the commits into live branch_)
* If PushArgs not set, then the script will fail. We have to take care of this case which not fail previously.

File: eng/common/pipelines/templates/steps/set-default-branch.yml (hunk 2 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/set-default-branch.yml:15]: Will have a pipeline on my private fork repo

Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/set-default-branch.yml]: https://dev.azure.com/azure-sdk/playground/_build/results?buildId=845015&view=logs&j=38d918fc-df5a-5870-2b80-6c17fdb13baf&t=489baed9-6f06-5266-9d0d-079a130738b4

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 1 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:44]: 1. That is has intended.
2. If you set the value to skip the check for changes then you are in charge of setting HasChanges to true if there are changes that are interesting. We want to avoid blindly pushing changes in a case where someone doesn't set HasChanges at all. 

As far as I'm aware this logic is working exactly has intended and if there is some scenario where you feel it is working please give me the specific case as I suspect that case needs to be address individually instead of changing the common template logic.

File: eng/common/pipelines/templates/steps/set-default-branch.yml (hunk 2 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/set-default-branch.yml]: Ok that looks better.

**Review by weshaggard — CHANGES_REQUESTED** [review]
The set-default-branch changes look good but I'd like to either undo or better understand why you need to change get-push-changes.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 1 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/git-push-changes.yml]: Totally agree with you as the changes here might breaking some feature. I removed it from this PR. Will open a new one if it is needed.

**Review by weshaggard — APPROVED** [review]
